framework/src/org/checkerframework/framework/util/ViewpointAdapter.java

```diff
@@ -0,0 +1,354 @@
+            Modifier resultModifier =
+                    combineModifierWithModifier(receiver, declModifier, atypeFactory);
+            aat.replaceAnnotation(extractAnnotationMirror(resultModifier));
+
+            // Combine component type recursively and sets combined component type
+            AnnotatedTypeMirror compo = aat.getComponentType();
+            // Recursively call itself first on the component type
+            AnnotatedTypeMirror combinedCompoType =
+                    combineModifierWithType(receiver, compo, atypeFactory);
+            aat.setComponentType(combinedCompoType);
+
+            return aat;
+        } else if (declared.getKind() == TypeKind.WILDCARD) {
+            AnnotatedWildcardType awt = (AnnotatedWildcardType) declared.shallowCopy();
+
+            Map<AnnotatedTypeMirror, AnnotatedTypeMirror> mapping = new HashMap<>();
+
+            // There is no main modifier for a wildcard
+
+            // Adapt extend
+            AnnotatedTypeMirror extend = awt.getExtendsBound();
+            if (extend != null) {
+                // Recursively adapt the extends bound of this awt
+                AnnotatedTypeMirror combinedExtend =
+                        combineModifierWithType(receiver, extend, atypeFactory);
+                mapping.put(extend, combinedExtend);
+            }
+
+            // Adapt super
+            AnnotatedTypeMirror zuper = awt.getSuperBound();
+            if (zuper != null) {
+                // Recursively adapt the lower bound of this awt
+                AnnotatedTypeMirror combinedZuper =
+                        combineModifierWithType(receiver, zuper, atypeFactory);
+                mapping.put(zuper, combinedZuper);
+            }
+
+            AnnotatedTypeMirror result = AnnotatedTypeReplacer.replace(awt, mapping);
+
+            return result;
+        } else if (declared.getKind() == TypeKind.NULL) {
+            AnnotatedNullType ant = (AnnotatedNullType) declared.shallowCopy(true);
+            Modifier declModifier = extractModifier(ant, atypeFactory);
+            Modifier result = combineModifierWithModifier(receiver, declModifier, atypeFactory);
+            ant.replaceAnnotation(extractAnnotationMirror(result));
+            return ant;
+        } else {
+            ErrorReporter.errorAbort(
+                    "ViewpointAdaptor::combineModifierWithType: Unknown decl: "
+                            + declared
+                            + " of kind: "
+                            + declared.getKind());
+            return null;
+        }
+    }
+
+    /**
+     * Extracts {@link AnnotationMirror} from a modifier.
+     *
+     * @param modifier Source modifier from which {@link AnnotationMirror} is extracted
+     * @return {@link AnnotationMirror} extracted
+     */
+    protected abstract AnnotationMirror extractAnnotationMirror(Modifier modifier);
+
+    /**
+     * Viewpoint adapt declared modifier to receiver modifier.
```

Comment:
Add "Subclass should override this method to provide viewpoint adaptation rules"